### PR TITLE
add for_buf function

### PR DESCRIPTION
**Pull request by secana**
Add a function that takes a buffer that is already in memory and scan it with vaas instead of a file path or stream.

### CODE REVIEW

**Review by GermanCoding — COMMENTED** [review]
Just a thought, without a concrete implementation, but perhaps it would be easier to write a conversion for slices to streams*, so that those can be used trivially with the existing for_stream() method?

The main downside compared to this approach would be that for_stream hashes server-side, so verdicts can be slower than hashing client-side. Though we could also keep for_buf, hash client-side and then upload buf as a stream.

File: rust/src/connection.rs [hunk 1]
```diff
@@ -189,7 +189,13 @@ impl Connection {
 
     /// Request a verdict for a file.
     pub async fn for_file(&self, file: &Path, ct: &CancellationToken) -> VResult<VaasVerdict> {
-        let sha256 = Sha256::try_from(file)?;
+        let buf = tokio::fs::read(file).await?;
```
[the remark this answers is not in the document]
Comment (secana) [reply rust/src/connection.rs:192]: I thought about that, too. I'm really not happy with the stream implementation, as it is very complicated and nearly unusable. That's why I added the simple buffer version. I'll have a look at the stream implementation and see if I can simplify it significantly. If that is possible I'll merge the two, if not I'll keep them separate.

[the remark this answers is not in the document]
Comment (secana) [reply rust/src/connection.rs:192]: > This loads the entire file into memory, which might consume a lot of RAM. The old code uses a streaming approach during uploading to upload large files efficiently.
> 
> (The SHA256 creation has always loaded the entire file into memory, but I wanted to refactor that at some point as well)

Does it matter? We have to load the whole file in any case into memory. Even if streamed, the whole file will be in memory at some point to be send over the network.

[the remark this answers is not in the document]
Comment (GermanCoding) [reply rust/src/connection.rs:192]: The method we use for uploading will locally discard bytes that have already been uploaded, so only a part ("chunk") of the file has to be kept in memory during upload.

**Comment by secana** [comment]
> I agree our current function signature for streams isn't great and should be improved. That said, it is possible to use bufs + for_stream with about 5 lines of code like this:
> 
> [8339801](https://github.com/GDATASoftwareAG/vaas/commit/8339801df46fad8fe00d68f1d8fc3f7384fdf476)
> 
> ```
>     pub async fn for_buf(&self, buf: Vec<u8>, ct: &CancellationToken) -> VResult<VaasVerdict> {
>         let len = buf.len();
>         let stream = stream::once(async move {
>             Ok::<Vec<u8>, std::io::Error>(buf)
>         });
>         self.for_stream(stream, len, ct).await
>     }
> ```

I like that, I'll change the code to use stream internally. What we have to decide is, if we want to break the API and use `Stream` in the `for_stream` function. Currently we use `TryStream` which make the API complicated, as everything has to be wrapped into a fail-able async type. We did that back then, 'cause that's the type that `reqwest` returns, which makes it plug'n'play with taking a `reqwest` response and putting it into a stream request. But I'm not sure if that is what the `for_stream` API should do. What do you think about that? 

Maybe we just add the `for_buf` method (wich uses `for_stream` internally) for a `6.1.0` version and think about a cleaner `for_stream` for a `7.x.x.` function, 'cause it's a breaking change, if we change the API.

**Comment by secana** [comment]
Okey, the `for_stream` does not check the SHA256, so it's not an option to use it in that case.

**Comment by GermanCoding** [comment]
I had yet another refactoring idea in #618: Generalize the uploading altogether, adding a new trait that specifies upload-data types. With this trait, adding support for things like for_buf (& many other types) is trivial with minimal code changes. The design still isn't perfect, but do you have an opinion on this?

**Pull request merged**